pkg/kube/client.go

```diff
@@ -373,19 +382,74 @@ func (c *client) RunAndWait(stop <-chan struct{}) {
 	c.metadataInformer.Start(stop)
 	c.istioInformer.Start(stop)
 	c.serviceapisInformers.Start(stop)
-	// WaitForCacheSync will virtually never be synced on the first call, as its called immediately after Start()
-	// This triggers a 100ms delay per call, which is often called 2-3 times in a test, delaying tests.
-	// Injecting an extremely small wait gives the tests some time to catch up in most cases, while having a trivial
-	// impact on non-test cases.
-	// Ideally, in the fake client we would just aggressively poll, but the informer neither exposes tuning for this
-	// nor gives us insight into which informers we need to wait for ourselves.
-	// https://github.com/kubernetes/kubernetes/issues/95262 tracks first class support
 	time.Sleep(time.Millisecond * 5)
-	c.kubeInformer.WaitForCacheSync(stop)
-	c.dynamicInformer.WaitForCacheSync(stop)
-	c.metadataInformer.WaitForCacheSync(stop)
-	c.istioInformer.WaitForCacheSync(stop)
-	c.serviceapisInformers.WaitForCacheSync(stop)
+	if c.fastSync {
+		// WaitForCacheSync will virtually never be synced on the first call, as its called immediately after Start()
+		// This triggers a 100ms delay per call, which is often called 2-3 times in a test, delaying tests.
+		// Instead, we add an aggressive sync polling
+		fastWaitForCacheSync(c.kubeInformer)
+		fastWaitForCacheSyncDynamic(c.dynamicInformer)
+		fastWaitForCacheSyncDynamic(c.metadataInformer)
+		fastWaitForCacheSync(c.istioInformer)
+		fastWaitForCacheSync(c.serviceapisInformers)
+	} else {
+		c.kubeInformer.WaitForCacheSync(stop)
+		c.dynamicInformer.WaitForCacheSync(stop)
+		c.metadataInformer.WaitForCacheSync(stop)
+		c.istioInformer.WaitForCacheSync(stop)
+		c.serviceapisInformers.WaitForCacheSync(stop)
+	}
+}
+
+type reflectInformerSync interface {
+	WaitForCacheSync(stopCh <-chan struct{}) map[reflect.Type]bool
+}
+
+type dynamicInformerSync interface {
+	WaitForCacheSync(stopCh <-chan struct{}) map[schema.GroupVersionResource]bool
+}
+
+// Wait for cache sync immediately, rather than with 100ms delay which slows tests
+// See https://github.com/kubernetes/kubernetes/issues/95262#issuecomment-703141573
+func fastWaitForCacheSync(informerFactory reflectInformerSync) {
+	returnImmediately := make(chan struct{})
+	close(returnImmediately)
+	_ = wait.PollImmediate(time.Microsecond, wait.ForeverTestTimeout, func() (bool, error) {
+		for _, synced := range informerFactory.WaitForCacheSync(returnImmediately) {
+			if !synced {
+				return false, nil
+			}
+		}
+		return true, nil
+	})
+}
+
+func fastWaitForCacheSyncDynamic(informerFactory dynamicInformerSync) {
```

Comment:
What's the difference with fastWaitForCacheSync

Reply:
Different signature, they don't have the same interface unfortunately. I need to wait for guidance on https://github.com/kubernetes/kubernetes/issues/95372 before this PR though or it will cause flakes